# Tree sequence workflow - tswf

[![Snakemake](https://img.shields.io/badge/snakemake-â‰¥7.0-brightgreen.svg)](https://snakemake.bitbucket.io)
[![Tests](https://github.com/percyfal/tswf/actions/workflows/main.yaml/badge.svg)][tests]
[![Codecov](https://codecov.io/gh/percyfal/tswf/branch/main/graph/badge.svg)][codecov]
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)][pre-commit]
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)][black]

[tests]: https://github.com/percyfal/tswf/actions?workflow=Tests
[codecov]: https://app.codecov.io/gh/percyfal/tswf
[pre-commit]: https://github.com/pre-commit/pre-commit
[black]: https://github.com/psf/black

## About

Python package consisting of snakemake workflow and utilities for
whole genome genealogy analyses, with a focus on tree sequencing data.
Supported methods include:

- [tsinfer](https://github.com/tskit-dev/tsinfer "tree sequence inference")
- [Relate](https://myersgroup.github.io/relate/ "estimate genome-wide genealogies")

## Features

### tsinfer

The workflow will generate tree sequences, calculate genealogical
nearest neighbours and produce interactive plots with
[bokeh](https://docs.bokeh.org/en/latest/).

## Installation

Easiest is to install with pip, possibly in a conda environment:

    mamba create -n tswf python=3.10
    python -m pip install git+https://github.com/percyfal/tswf

### Dependencies

Until there is a conda package, you need to install the dependencies
manually.

    mamba install -c conda-forge -c bioconda -c default  snakemake-minimal bcftools bokeh=2.4.3 scipy matplotlib

## Quickstart

Run

    tswf quickstart

for information on how to initialize an analysis, setup configuration
files and more.

## Usage

The workflow and additional commands run via the main entry point:

    tswf --help
    tswf config example workflow
    tswf smk run -j 1
    tswf smk run -j 1 --use-envmodules --envmodules-file=envmodules.yaml

See the subcommand help for more information.

## Authors

- Per Unneberg (@percyfal)

## Configuration

tswf requires several configuration files to run. All configuration
files are defined by and validated against schemas. The main `tswf`
configuration file resides in the top-level project directory. You can
generate example configuration file with the command `tswf config
example` or generate it with `tswf config init`. An example
configuration file looks as follows:

    project_name: tswf
    snakemake_profiles:
      local: config/local

### Workflow configuration

The snakemake workflow requires a configuration file, by default
residing in `config/config.yaml` relative to the working directory,
that defines location of samplesheet, population data sheet, what
samples and analyses to run, and more. Use the command `tswf config
example workflow` to generate an example.

### Samples and population data sheets

The samples and population data sheets are tab-separated text files
that define input samples and populations. The samples file requires
two columns `SM` and `population`:

    SM      population
    tsk_1   CHB
    tsk_7   CEU
    tsk_19  chimp
    tsk_20  gorilla
    tsk_21  orangutan

For more options see the example generated by `tswf config example
samples`.

Similarly, the population data sheet requires columns `population` and
`species`:

    population   species
    CHB    Homo sapiens
    CEU    Homo sapiens
    chimp     Pan troglodytes

For more options see the example generated by `tswf config example
populations`.

### Environment module configuration

If the workflow is run on a HPC with the `--use-envmodules` option
(see
[using-environment-modules](https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#using-environment-modules)),
the workflow will check for an additional configuration file that
configures environment modules. By default, the file is
`config/envmodules.yaml`, but a custom location can be set with the
environment variable `TSWF_ENVMODULES` or the `--envmodules-file`
option.

envmodules configurations are placed in a configuration section
`envmodules` with key-value pairs that map a dependency set to a list
of environment modules. The dependency sets are named after the rule's
corresponding conda environment file, such that a dependency set may
affect multiple rules. For instance, the following example shows how
to define modules for rules depending on fastqc, as it would be
implemented on the [uppmax](https://uppmax.uu.se/) compute cluster:

    envmodules:
      fastqc:
        - bioinfo-tools
        - FastQC

See the configuration schema file
(`workflow/schemas/config.schema.yaml`) or the example generated with
`tswf config example envmodules` for more information.

## Scripts

There are several tools that are run by the workflow that are also
available as standalone scripts. All scripts are prefixed with
`tswf-`, and available scripts can be found by pressing `tswf-`
followed by `<TAB>`. For instance, `tswf-get-ancestral-allele` is a
script for inferring ancestral alleles in a vcf file.

## Testing

Test cases are in the subfolder `tests` in the source distribution and
included in the installation. They are automatically executed via
continuous integration with [Github
Actions](https://github.com/features/actions).

Tests can be run locally by issuing

    tswf smk run --test --testdir test -j 1

The `--testdir` option sets up test files in a directory of choice;
else, the tests are run in a `/tmp` folder.
